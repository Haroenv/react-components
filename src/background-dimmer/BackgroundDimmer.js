import React from 'react';

/**
 * ### Creates a full-page dimmed background for its children nodes
 */
const BackgroundDimmer = React.createClass({

  propTypes: {
    /**
     * children nodes/elements
     */
    children: React.PropTypes.node.isRequired,
    /**
     * background-color
     */
    color: React.PropTypes.string,
    /**
     * opacity
     */
    alpha: React.PropTypes.number,
    /**
     * avoid propagation for scroll events
     */
    stopScrollPropagation: React.PropTypes.bool,
    /**
     * called when user clicks outside children
     */
    onClickOutside: React.PropTypes.func,
    className: React.PropTypes.string,
    id: React.PropTypes.string,
    style: React.PropTypes.object
  },

  getDefaultProps() {
    return {
      color: 'black',
      alpha: 0.5
    };
  },

  componentDidMount() {
    if (this.props.stopScrollPropagation) {
      this.disableScrollPropagation();
    }
  },

  enableScrollPropagation() {
    const dimmedBackground = this.refs.dimmedBackground.getDOMNode();
    dimmedBackground.removeEventListener('wheel', this.stopScrollPropagation);
    dimmedBackground.removeEventListener('touchmove', this.stopScrollPropagation);
  },

  disableScrollPropagation() {
    const dimmedBackground = this.refs.dimmedBackground.getDOMNode();
    dimmedBackground.addEventListener('wheel', this.stopScrollPropagation);
    dimmedBackground.addEventListener('touchmove', this.stopScrollPropagation);
  },

  isEventOutsideChildren(e) {
    const el = e.target || e.srcElement;
    return el === this.refs.dimmedBackground.getDOMNode();
  },

  onClick(e) {
    const { onClickOutside } = this.props;
    if (onClickOutside && this.isEventOutsideChildren(e)) {
      onClickOutside(e);
    }
  },

  stopScrollPropagation(e) {
    const { stopScrollPropagation } = this.props;
    if (stopScrollPropagation && this.isEventOutsideChildren(e)) {
      e.preventDefault();
    }
  },

  getDimmedBackground() {
    const { color, alpha } = this.props;
    const style = {
      position: 'fixed',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      backgroundColor: color,
      opacity: String(alpha)
    };
    return <div style={style} onClick={this.onClick} ref='dimmedBackground' />;
  },

  render() {
    const { style, className, id, children } = this.props;
    const mergedStyle = {
      position: 'relative',
      ...style
    };
    return (
      <div {...{ style: mergedStyle, className, id }}>
        {this.getDimmedBackground()}
        <div style={{ visibility: 'hidden' }}>{children}</div>
        <div style={{ position: 'absolute', width: '100%', height: '100%', top: 0, left: 0 }}>
          {children}
        </div>
      </div>
    );
  },

  componentWillReceiveProps(nextProps) {
    if (!nextProps.stopScrollPropagation && this.props.stopScrollPropagation) {
      this.enableScrollPropagation();
    } else if (nextProps.stopScrollPropagation && !this.props.stopScrollPropagation) {
      this.disableScrollPropagation();
    }
  },

  componentWillUnmount() {
    if (this.props.stopScrollPropagation) {
      this.enableScrollPropagation();
    }
  }

});

export default BackgroundDimmer;
