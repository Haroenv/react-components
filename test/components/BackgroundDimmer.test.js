import React from 'react';
import renderer from 'react-test-renderer';
// Because of this bug: https://github.com/facebook/react/issues/7386
// Should be fixed in react 15.4
// import { shallow } from 'enzyme';

import BackgroundDimmer from '../../src/background-dimmer';

// Because of this bug: https://github.com/facebook/react/issues/7386
// Should be fixed in react 15.4
jest.mock('react-dom');

const content = <div className='content'>content</div>;

describe('BackgroundDimmer', () => {

  it('renders correctly', () => {
    const component = renderer.create(
      <BackgroundDimmer>
        {content}
      </BackgroundDimmer>
    );
    expect(component).toMatchSnapshot();
  });

  // Ignored because of this bug: https://github.com/facebook/react/issues/7386
  // Should be fixed in react 15.4
  xit('invokes the callback when the background is clicked', () => {
    const onClickOutside = jest.fn();
    const dimmer = shallow( // eslint-disable-line no-undef
      <BackgroundDimmer onClickOutside={onClickOutside}>
        {content}
      </BackgroundDimmer>
    );
    dimmer.find('.main-content-wrapper').simulate('click');
    expect(onClickOutside).toBeCalled();
  });

  // Ignored because of this bug: https://github.com/facebook/react/issues/7386
  // Should be fixed in react 15.4
  xit('doesn\'t invoke the callback when the content is clicked', () => {
    const onClickOutside = jest.fn();
    const dimmer = shallow( // eslint-disable-line no-undef
      <BackgroundDimmer onClickOutside={onClickOutside}>
        {content}
      </BackgroundDimmer>
    );
    dimmer.find('.content').simulate('click');
    expect(onClickOutside).not.toBeCalled();
  });

  describe('getLocals', () => {

    it('uses default props', () => {
      const dimmer = new BackgroundDimmer({
        children: content
      });
      const locals = dimmer.getLocals();
      expect(locals).toMatchSnapshot();
    });

  });

});
