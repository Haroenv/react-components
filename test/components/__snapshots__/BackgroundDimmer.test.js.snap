exports[`BackgroundDimmer getLocals uses default props 1`] = `
Object {
  "centeredContentWrapperProps": Object {
    "className": "centered-content-wrapper",
    "column": true,
    "onClick": [Function anonymous],
    "style": Object {
      "height": "auto",
      "maxHeight": "90%",
      "maxWidth": "90%",
      "width": "auto"
    }
  },
  "children": <div
    className="content">
    content
</div>,
  "className": "background-dimmer",
  "mainContentWrapperProps": Object {
    "className": "main-content-wrapper",
    "hAlignContent": "center",
    "onClick": [Function anonymous],
    "onTouchMove": [Function anonymous],
    "onWheel": [Function anonymous],
    "ref": "mainContentWrapper",
    "style": Object {
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
      "zIndex": 100000
    },
    "vAlignContent": "center"
  },
  "overlayProps": Object {
    "style": Object {
      "backgroundColor": "black",
      "bottom": 0,
      "left": 0,
      "opacity": "0.5",
      "position": "fixed",
      "right": 0,
      "top": 0,
      "zIndex": 99999
    }
  },
  "stopPropagation": [Function anonymous]
}
`;

exports[`BackgroundDimmer renders correctly 1`] = `
<div
  className="background-dimmer"
  stopPropagation={[Function anonymous]}>
  <div
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "left": 0,
        "opacity": "0.5",
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 99999
      }
    } />
  <div
    className="react-flex-view align-content-center justify-content-center main-content-wrapper"
    onClick={[Function anonymous]}
    onTouchMove={[Function anonymous]}
    onWheel={[Function anonymous]}
    style={
      Object {
        "MozBoxFlex": "0 1 auto",
        "WebkitBoxFlex": "0 1 auto",
        "WebkitFlex": "0 1 auto",
        "bottom": 0,
        "flex": "0 1 auto",
        "left": 0,
        "msFlex": "0 1 auto",
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 100000
      }
    }>
    <div
      className="react-flex-view flex-column centered-content-wrapper"
      onClick={[Function anonymous]}
      style={
        Object {
          "MozBoxFlex": "0 1 auto",
          "WebkitBoxFlex": "0 1 auto",
          "WebkitFlex": "0 1 auto",
          "flex": "0 1 auto",
          "height": "auto",
          "maxHeight": "90%",
          "maxWidth": "90%",
          "msFlex": "0 1 auto",
          "width": "auto"
        }
      }>
      <div
        className="content">
        content
      </div>
    </div>
  </div>
</div>
`;
